### ISSUE TITLE
nn.Linear on CUDA accepts incorrect input shape

### ISSUE THREAD

**Issue by DianChen-TRI** [pytorch/pytorch#81659]
### 🐛 Describe the bug

The `nn.Linear` on CUDA accepts incorrect input shapes that **don't fit** the `in_channels` and weirdly ouputs a tensor with the expected shape. It's really confusing since I don't even know which column to inspect to check the mathematical calculation.

sample code:
```python
>>> import torch
# CPU version works as expected
>>> f = torch.nn.Linear(10, 1024)
>>> a = torch.randn(1, 6, 32, 88, 192)
>>> f(a).shape
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 96, in forward
    return F.linear(input, self.weight, self.bias)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py", line 1847, in linear
    return torch._C._nn.linear(input, weight, bias)
RuntimeError: mat1 and mat2 shapes cannot be multiplied (16896x192 and 10x1024)
# CUDA version has weird behavior
>>> f = f.to('cuda')
>>> a = a.cuda()
>>> f(a).shape
torch.Size([1, 6, 32, 88, 1024])
# remove a dimension, still happens
>>> a = torch.randn(1, 6, 32, 192)
>>> a = a.cuda()
>>> f(a).shape
torch.Size([1, 6, 32, 1024])
# remove a dimension, still happens
>>> a = torch.randn(1, 6, 192)
>>> a = a.cuda()
>>> f(a).shape
torch.Size([1, 6, 1024])
# happens in an unpredictable way
>>> a = torch.randn(2, 6, 5)
>>> a = a.cuda()
>>> f(a).shape
 ** On entry to SGEMM  parameter number 10 had an illegal value
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 96, in forward
    return F.linear(input, self.weight, self.bias)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py", line 1847, in linear
    return torch._C._nn.linear(input, weight, bias)
RuntimeError: CUDA error: CUBLAS_STATUS_INVALID_VALUE when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`
>>> a = torch.randn(2, 6, 4)
>>> a = a.cuda()
>>> f(a).shape
 ** On entry to SGEMM  parameter number 10 had an illegal value
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1051, in _call_impl
    return forward_call(*input, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/linear.py", line 96, in forward
    return F.linear(input, self.weight, self.bias)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py", line 1847, in linear
    return torch._C._nn.linear(input, weight, bias)
RuntimeError: CUDA error: CUBLAS_STATUS_INVALID_VALUE when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`
>>> a = torch.randn(2, 6, 192)
>>> a = a.cuda()
>>> f(a).shape
torch.Size([2, 6, 1024])
```

### Versions

Collecting environment information...
PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.10

Python version: 3.7.10 (default, Feb 26 2021, 18:47:35)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-1030-aws-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: A100-SXM4-40GB
GPU 1: A100-SXM4-40GB
GPU 2: A100-SXM4-40GB
GPU 3: A100-SXM4-40GB
GPU 4: A100-SXM4-40GB
GPU 5: A100-SXM4-40GB
GPU 6: A100-SXM4-40GB
GPU 7: A100-SXM4-40GB

Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.19.5
[pip3] pytorch3d==0.3.0
[pip3] torch==1.9.0
[pip3] torchelastic==0.2.0
[pip3] torchtext==0.10.0
[pip3] torchvision==0.10.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.1.74              h6bb024c_0    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-service               2.3.0            py37h27cfd23_1  
[conda] mkl_fft                   1.3.0            py37h42c9631_2  
[conda] mkl_random                1.2.1            py37ha9443f7_2  
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] pytorch                   1.9.0           py3.7_cuda11.1_cudnn8.0.5_0    pytorch
[conda] pytorch3d                 0.3.0                    pypi_0    pypi
[conda] torchelastic              0.2.0                    pypi_0    pypi
[conda] torchtext                 0.10.0                     py37    pytorch
[conda] torchvision               0.10.0               py37_cu111    pytorch

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @ngimel

**Comment by gchanan** [comment]
can you check 1.12?  I recall this being fixed at some point.

**Comment by ngimel** [comment]
This was fixed for 1.9.1 and later, please reopen if that's not the case.

**Comment by cspzyy** [comment]
Hi, it seems an issue long time ago. But I met the totally same trouble now.
Any progress about this issue, please? QAQ
